packages/edge-preview-authenticated-proxy/src/index.ts

```diff
@@ -0,0 +1,328 @@
+			"Access-Control-Allow-Method": "POST",
+		},
+	});
+}
+
+async function handleRequest(request: Request, ctx: ExecutionContext) {
+	const url = new URL(request.url);
+
+	/**
+	 * Request the preview session associated with a given exchange_url
+	 * exchange_url comes from an authenticated core API call made in the client
+	 * It doesn't have CORS set up, so needs to be proxied
+	 */
+	if (
+		request.method === "POST" &&
+		url.hostname === "preview.devprod.cloudflare.dev" &&
+		url.pathname === "/exchange"
+	) {
+		return handleTokenExchange(url);
+	}
+
+	/**
+	 * Given a preview session, the client should obtain a specific preview token
+	 * This endpoint takes in the URL parameters:
+	 * 	- `token`   The preview token to authenticate future preview requests
+	 * 						  Crucially, this is _different_ from the session token obtained above
+	 *  - `remote`  Which endpoint to hit with preview requests
+	 *              This should be the workers.dev deployment for the current worker
+	 *  - `prewarm` A fire-and-forget prewarm endpoint to hit to start up the preview
+	 *  - `suffix`  (optional) The pathname + search to hit on the preview worker once redirected
+	 *
+	 * It must be called with a random subdomain (i.e. some-random-data.preview.devprod.cloudflare.dev)
+	 * to provide cookie isolation for the preview.
+	 *
+	 * It will redirect to the suffix provide, setting a cookie with the `token` and `remote`
+	 * for future use.
+	 */
+	if (
+		request.method === "GET" &&
+		url.hostname.endsWith("preview.devprod.cloudflare.dev") &&
+		url.pathname === "/.update-preview-token"
+	) {
+		return updatePreviewToken(url, ctx);
+	}
+
+	/**
+	 * Given a preview token, this endpoint allows for raw http calls to be inspected
+	 * It must be called with a random subdomain (i.e. some-random-data.rawhttp.devprod.cloudflare.dev)
+	 * for consistency with the preview endpoint. This is not currently used, but may be in future
+	 *
+	 * It required two parameters, passed as headers:
+	 *  - `X-CF-Token`  A preview token, as in /.update-preview-token
+	 *  - `X-CF-Remote` Which endpoint to hit with preview requests, as in /.update-preview-token
+	 */
+	if (url.hostname.endsWith("rawhttp.devprod.cloudflare.dev")) {
+		return handleRawHttp(request, url);
+	}
+
+	/**
+	 * Finally, if no other conditions are met, make a preview request to the worker
+	 * This must be called with the same subdomain as used in /.update-preview-token
+	 * so that the cookie will be present. It will swap the host and inject the preview token
+	 * but otherwise will pass the request through unchanged
+	 */
+	const parsedCookies = cookie.parse(request.headers.get("Cookie") ?? "");
+	const { token, remote } = JSON.parse(parsedCookies?.token ?? "{}");
+	if (!token || !remote) {
+		throw new PreviewRequestFailed();
+	}
+
+	const workerUrl = new URL(url);
+	const remoteUrl = new URL(remote);
+	workerUrl.hostname = remoteUrl.hostname;
+	workerUrl.protocol = remoteUrl.protocol;
```

Comment:
NIT: This code is the same as elsewhere in the file. Perhaps a helper function to avoid the duplication?

Reply:
Same question about ports.